### DISCUSSION TITLE
textcat_multilabel using accuracy model

### DISCUSSION THREAD

**Discussion by lmanuelfigueroa** [explosion/spaCy#8880] [Help: Coding & Implementations] [answered]
Hi I am looking for some help with the textcat_multilabel pipeline using accuracy model.  I am training a model on 3 labels "war", "school", and "work". During the training the scores peak and then drop and stabilize. Looking at the meta.json file from the model all labels have a score of 0. Results from the training are on the image below.  
<img width="711" alt="results" src="https://user-images.githubusercontent.com/16822920/128098999-646a225b-1264-42c2-8427-b772055fcf85.PNG">
<img width="250" alt="json" src="https://user-images.githubusercontent.com/16822920/128099006-5065f1c3-7bbe-407e-8a2f-92c6a8eb7384.PNG">

I need some guidance on what could be the issue. I have been able to successfully run the model in efficiency with good results, on the same dataset which leads me to believe I might be missing something when training with the accuracy model. Or it has something to do with the RunningTimeWarning or with the dataset not being useful for the accuracy model. I am using Windows10,spacy3.1.1 and I am using the base_config.cfg for accuracy on CPU from [https://spacy.io/usage/training#config](url) .  Here is the code for the program. 

I appreciate any feedback. 

```
def train_model() -> None:
    print("Training Model ....")

    nlp = spacy.load("en_core_web_lg")
    
    training_data = []
    validate_data = []

    labels = ["work","war","school"]
   
    # creating the training samples for model with labels
    for label in labels:
        
        file = open("model_data/"+ label +"/train.txt","r",encoding = "utf-8")

        for line in file:
            training_data.append((line,label))
        file.close()

    #randomizing the data
    training_data = random.sample(training_data,len(training_data))

    #creating the valid samples for model with labels

    for label in labels:
        file = open("model_data/" + label + "/valid.txt","r",encoding = "utf-8")

        for line in file:
            validate_data.append((line,label))
        file.close()


    #randomizing the data
    validate_data = random.sample(validate_data,len(validate_data))

   
    #create the model 
    train_docs = make_docs(nlp,training_data)
    doc_bin = DocBin(docs=train_docs)
    
    doc_bin.to_disk("./train_acc.spacy")

    valid_docs = make_docs(nlp,validate_data)

    doc_bin = DocBin(docs=valid_docs)
   
    doc_bin.to_disk("./valid_acc.spacy")
    

    print("Done")




def make_docs(nlp, data):



    docs = []
    for doc, label in nlp.pipe(data, as_tuples=True):

        doc.cats["war"] = 0
        doc.cats["work"] = 0
        doc.cats["school"] = 0


        if label == "war":
            doc.cats["war"] = 1

        elif label == "work":
            doc.cats["work"] = 1
        
        elif label == "school":
            doc.cats["school"] = 1

        docs.append(doc)
    return (docs) 


def test_model() -> None:

    TESTING_SAMPLES = 3

    nlp = spacy.load("./model-last")

    tags = ["work","war","school"]

    for tag in tags:
        print("\n===========================================================================\n")

        tag_file = open("./model_data/"+ tag + "/test.txt","r",encoding = "utf-8")
    
        print("TESTING " + tag.upper() + " CATEGORY\n")

        text = []
        for line in tag_file:
            text.append(line)
        print("\n")

        for i in range(TESTING_SAMPLES):
            print(text[i].strip())
            doc = nlp(text[i])
            print(doc.cats)
            #print_top_labels(doc.cats)
            print("\n")
        tag_file.close()
        

if __name__ == "__main__":


    try:
        if sys.argv[1] == "train":
            train_model()
        elif sys.argv[1] == "test":
            test_model()
    except IndexError:
        print("use \" python multilcass.py train\" to train a model or \" python multiclass.py test\" to test the model")
```

**Comment by polm** [comment, marked as answer]
Please don't share screenshots of text/code, please copy/paste the text data.

You're using textcat multilabel but it doesn't look like you actually have multilabel documents. Multilabel is like if you have one document that is both "school" and "war". But the way your if statement is set up it's not possible for a document to have multiple labels, though it is possible for a document to have no labels. Do you have documents with no labels?

Can you check that your working config has the same settings regarding exclusive classes?

It would be great if you could share both your configs (as text) and the output of `spacy debug data` (see [docs](https://spacy.io/api/cli#debug-data)). 

The `RuntimeWarning` should not be an issue, though I'll link to https://github.com/explosion/spaCy/discussions/8049 where it came up before.

**Reply by lmanuelfigueroa** [reply]
Thank you Paul for taking the time. 

**This is the base_config.cfg that I used from spacy website** 
`[paths]
train = "./model_data/train.spacy"
dev = "./model_data/valid.spacy"

[system]
gpu_allocator = null

[nlp]
lang = "en"
pipeline = ["tok2vec","textcat_multilabel"]
batch_size = 1000

[components]

[components.tok2vec]
factory = "tok2vec"

[components.tok2vec.model]
@architectures = "spacy.Tok2Vec.v2"

[components.tok2vec.model.embed]
@architectures = "spacy.MultiHashEmbed.v2"
width = ${components.tok2vec.model.encode.width}
attrs = ["ORTH", "SHAPE"]
rows = [5000, 2500]
include_static_vectors = true

[components.tok2vec.model.encode]
@architectures = "spacy.MaxoutWindowEncoder.v2"
width = 256
depth = 8
window_size = 1
maxout_pieces = 3

[components.textcat_multilabel]
factory = "textcat_multilabel"

[components.textcat_multilabel.model]
@architectures = "spacy.TextCatEnsemble.v2"
nO = null

[components.textcat_multilabel.model.tok2vec]
@architectures = "spacy.Tok2VecListener.v1"
width = ${components.tok2vec.model.encode.width}

[components.textcat_multilabel.model.linear_model]
@architectures = "spacy.TextCatBOW.v2"
exclusive_classes = false
ngram_size = 1
no_output_layer = false

[corpora]

[corpora.train]
@readers = "spacy.Corpus.v1"
path = ${paths.train}
max_length = 0

[corpora.dev]
@readers = "spacy.Corpus.v1"
path = ${paths.dev}
max_length = 0

[training]
dev_corpus = "corpora.dev"
train_corpus = "corpora.train"

[training.optimizer]
@optimizers = "Adam.v1"

[training.batcher]
@batchers = "spacy.batch_by_words.v1"
discard_oversize = false
tolerance = 0.2

[training.batcher.size]
@schedules = "compounding.v1"
start = 100
stop = 1000
compound = 1.001

[initialize]
vectors = "en_core_web_lg"`

**This is the config.cfg file generated** 

`[paths]
train = "./model_data/train.spacy"
dev = "./model_data/valid.spacy"
vectors = null
init_tok2vec = null

[system]
gpu_allocator = null
seed = 0

[nlp]
lang = "en"
pipeline = ["tok2vec","textcat_multilabel"]
batch_size = 1000
disabled = []
before_creation = null
after_creation = null
after_pipeline_creation = null
tokenizer = {"@tokenizers":"spacy.Tokenizer.v1"}

[components]

[components.textcat_multilabel]
factory = "textcat_multilabel"
threshold = 0.5

[components.textcat_multilabel.model]
@architectures = "spacy.TextCatEnsemble.v2"
nO = null

[components.textcat_multilabel.model.linear_model]
@architectures = "spacy.TextCatBOW.v2"
exclusive_classes = false
ngram_size = 1
no_output_layer = false
nO = null

[components.textcat_multilabel.model.tok2vec]
@architectures = "spacy.Tok2VecListener.v1"
width = ${components.tok2vec.model.encode.width}
upstream = "*"

[components.tok2vec]
factory = "tok2vec"

[components.tok2vec.model]
@architectures = "spacy.Tok2Vec.v2"

[components.tok2vec.model.embed]
@architectures = "spacy.MultiHashEmbed.v2"
width = ${components.tok2vec.model.encode.width}
attrs = ["ORTH","SHAPE"]
rows = [5000,2500]
include_static_vectors = true

[components.tok2vec.model.encode]
@architectures = "spacy.MaxoutWindowEncoder.v2"
width = 256
depth = 8
window_size = 1
maxout_pieces = 3

[corpora]

[corpora.dev]
@readers = "spacy.Corpus.v1"
path = ${paths.dev}
max_length = 0
gold_preproc = false
limit = 0
augmenter = null

[corpora.train]
@readers = "spacy.Corpus.v1"
path = ${paths.train}
max_length = 0
gold_preproc = false
limit = 0
augmenter = null

[training]
dev_corpus = "corpora.dev"
train_corpus = "corpora.train"
seed = ${system.seed}
gpu_allocator = ${system.gpu_allocator}
dropout = 0.1
accumulate_gradient = 1
patience = 1600
max_epochs = 0
max_steps = 20000
eval_frequency = 200
frozen_components = []
annotating_components = []
before_to_disk = null

[training.batcher]
@batchers = "spacy.batch_by_words.v1"
discard_oversize = false
tolerance = 0.2
get_length = null

[training.batcher.size]
@schedules = "compounding.v1"
start = 100
stop = 1000
compound = 1.001
t = 0.0

[training.logger]
@loggers = "spacy.ConsoleLogger.v1"
progress_bar = false

[training.optimizer]
@optimizers = "Adam.v1"
beta1 = 0.9
beta2 = 0.999
L2_is_weight_decay = true
L2 = 0.01
grad_clip = 1.0
use_averages = false
eps = 0.00000001
learn_rate = 0.001

[training.score_weights]
cats_score = 1.0
cats_score_desc = null
cats_micro_p = null
cats_micro_r = null
cats_micro_f = null
cats_macro_p = null
cats_macro_r = null
cats_macro_f = null
cats_macro_auc = null
cats_f_per_type = null
cats_macro_auc_per_type = null

[pretraining]

[initialize]
vectors = "en_core_web_lg"
init_tok2vec = ${paths.init_tok2vec}
vocab_data = null
lookups = null
before_init = null
after_init = null

[initialize.components]

[initialize.tokenizer]`



**Here is the debug data** 
============================ Data file validation ============================
✔ Corpus is loadable
✔ Pipeline can be initialized with data

=============================== Training stats ===============================
Language: en
Training pipeline: tok2vec, textcat_multilabel
450 training docs
117 evaluation docs
✔ No overlap between training and evaluation data
⚠ Low number of examples to train a new pipeline (450)

============================== Vocab & Vectors ==============================
ℹ 14790 total word(s) in the data (2189 unique)
ℹ 684830 vectors (684830 unique keys, 300 dimensions)
⚠ 1615 words in training data without vectors (11%)

====================== Text Classification (Multilabel) ======================
ℹ Text Classification: 3 label(s)
⚠ The train data contains only instances with mutually-exclusive classes. You
can potentially use the component 'textcat' instead of 'textcat_multilabel'.

================================== Summary ==================================
✔ 3 checks passed
⚠ 3 warnings

**Reply by lmanuelfigueroa** [reply]
Currently all text have a label associated with them. My thinking was that if I have 3 labels "war" "work" and "fear". During training I would have a text like 

("But then the war started  I start to go in June in the air force  and in May the war started  and that was it.","war") 

then assigned it like 
doc.cats["war"] = 1 
doc.cats["fear"] = 0 
doc.cats["work"] = 0


("And it was the night when I decide I better go 18 to the airport to work.","work") 

then assigned it like 
doc.cats["war"] = 0 
doc.cats["fear"] = 0 
doc.cats["work"] = 1 


("And believe it or not  my father was afraid of wild animals.", "fear")

then assigned it like 
doc.cats["war"] = 0 
doc.cats["fear"] = 1 
doc.cats["work"] = 0

Then when testing a new text "The prison director said to people like my mother  you can go,  the war is over,  we are not afraid of spies anymore  you can leave." I would expect to receive let's say 50% war 50% fear and 0% work.

Is this the incorrect approach for multilabel and that is why the warning appears that train data contains only mutually-exclusive classes, and I should have training data where there is overlap ex:("The prison director said to people like my mother  you can go,  the war is over,  we are not afraid of spies anymore  you can leave.", "fear","war")

then assigned it
doc.cats["war"] = 1
doc.cats["fear"] = 1 
doc.cats["work"] = 0

**Reply by polm** [reply]
> Then when testing a new text "The prison director said to people like my mother you can go, the war is over, we are not afraid of spies anymore you can leave." I would expect to receive let's say 50% war 50% fear and 0% work.

That's not how the multilabel model works. It decides each label individually, so there's no guarantee the labels will add up to 100%. You can think of it as making a binary decision for each label of whether to apply that label or not. 

In general your training data should be as much like your production data as possible. If you want your model to learn how to combine labels it's possible it'll learn that without any examples, but I wouldn't bet on it.

**Reply by lmanuelfigueroa** [reply]
Okay, thank you for your help.
